File: 1539-kth-missing-positive-number/1539-kth-missing-positive-number.java
class Solution {
    public int findKthPositive(int[] arr, int k) {
        int i = 0;
        while (i < arr.length){
            int j = arr[i] - 1;
            if (arr[i] <= arr.length && arr[i] != arr[j])
                swap(arr, i, j);
            else
                i++;
        }
        List<Integer> missingNumbers = new ArrayList<>();
        Set<Integer> extraNumbers = new HashSet<>();
        for (i = 0; i < arr.length && missingNumbers.size() < k; i++){
            if (arr[i] != i+1){
                missingNumbers.add(i+1);
                extraNumbers.add(arr[i]);
            }
        }
        
        for (i = 1; missingNumbers.size() < k; i++){
            int candidate = i + arr.length;
            if (!extraNumbers.contains(candidate))
                missingNumbers.add(candidate);
        }
        
        return missingNumbers.get(k-1);
    }
    
    private void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: 1539-kth-missing-positive-number/1539-kth-missing-positive-number.py
class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        missing_arr = []
        for i in range(len(arr)):
            if i == 0:
                missing = arr[i] - 0 - 1
                for i in range(1, arr[i]):
                    missing_arr.append(i)
            else:
                missing = arr[i] - arr[i-1] - 1
                for i in range(arr[i-1] +1, arr[i]):
                    missing_arr.append(i)
            if len(missing_arr) >= k:
                print(missing_arr)
                return missing_arr[k-1]
       
        
        a = arr[-1] + 1
        while True:
            missing_arr.append(a)
            a += 1
            if len(missing_arr) >= k:
                print(missing_arr)
                return missing_arr[k-1]
        
            